File: JAVA/MinimumCostToConvertStringI.java
/**
 * @version JAVA
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package JAVA;

import java.util.Arrays;

public class MinimumCostToConvertStringI {

    public static void main(String[] args) {
        System.out.printf("Resultado: %d\n", minimumCost( "abcd", "acbe", new char[]{'a','b','c','c','e','d'}, new char[]{'b','c','b','e','b','e'}, new int[]{2,5,5,1,2,20} ));
    }

    public static long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
        int inf = Integer.MAX_VALUE / 2;
        int n = 26;
        int[][][] dp = new int[n + 1][n][n];

        for (int i = 0; i <= n; i++) {
            for (int j = 0; j < n; j++) {
                Arrays.fill(dp[i][j], inf);
            }
        }

        int[][] g = new int[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(g[i], inf);
            g[i][i] = 0;
        }

        for (int i = 0; i < original.length; i++) {
            int x = original[i] - 'a';
            int y = changed[i] - 'a';
            g[x][y] = Math.min(g[x][y], cost[i]);
        }

        dp[0] = g;
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    dp[k + 1][i][j] = Math.min(dp[k][i][j], dp[k][i][k] + dp[k][k][j]);
                }
            }
        }

        long ans = 0;
        for (int i = 0; i < source.length(); i++) {
            int a = source.charAt(i) - 'a';
            int b = target.charAt(i) - 'a';
            if (dp[n][a][b] >= inf) {
                return -1;
            }
            ans += dp[n][a][b];
        }
        return ans;
    }
}

File: GO/MinimumCostToConvertStringI.go
/**
 * @version GO 1.22.3
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package GO

import "math"

func minimumCost(source string, target string, original []byte, changed []byte, cost []int) int64 {
	inf := math.MaxInt / 2
	n := 26
	dp := make([][][]int, n+1)
	for i := range dp {
		dp[i] = make([][]int, n)
		for j := range dp[i] {
			dp[i][j] = make([]int, n)
		}
	}

	g := make([][]int, n)
	for i := range g {
		g[i] = make([]int, n)
		for j := range g[i] {
			if i != j {
				g[i][j] = inf
			}
		}
	}

	for i := 0; i < len(original); i++ {
		x, y, z := original[i]-'a', changed[i]-'a', cost[i]
		g[x][y] = min(g[x][y], z)
	}

	dp[0] = g
	for k := 0; k < n; k++ {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				dp[k+1][i][j] = min(dp[k][i][j], dp[k][i][k]+dp[k][k][j])
			}
		}
	}
	ans := 0
	for i := 0; i < len(source); i++ {
		a, b := source[i]-'a', target[i]-'a'
		if dp[n][a][b] >= inf {
			return -1
		}
		ans += dp[n][a][b]
	}
	return int64(ans)
}
